### PR TITLE
Canonify image URL for Twitter

### LINKED ISSUE #773: [Bug] Image payload cannot be decrypted on twitter
# Bug Report

### System

- OS:
- OS Version:

### Browser

- Browser: Any
- Browser Version:

### Maskbook

- Maskbook Version: Latest store ver
- Installation: /* May be "Store", "ZIP", or "Self-Complied" */
- Build Commit: Optionally attach a Commit ID, if it is from an pre-release branch head

### Expected Behavior

/* Write the expected behavior here. */

### Actual Behavior
https://dimension.chat/group/maskbook?msg=sN3qAauwbwEcTxuAQ

Cannot recognise the image payload
/* Write the actual behavior here. */

### How To Reproduce

Post a red packet  (or anything with image payload) on twitter

### PR DESCRIPTION
**Pull request by guanbinrui**
## Why Twitter image payload breaks?

Twitter may delivery different image URLs according to network condition or user preferred configrations. Some of them are highly compressed which can not be decoded.
 
## How to fix?

The URL of Twitter photo adhere [fixed formatting](https://developer.twitter.com/en/docs/tweets/data-dictionary/overview/entities-object#photo_format). We can get original photo with URL like `https://pbs.twimg.com/media/[id]?format=[jpg|png]&name=orig`.

## Refs

- https://stackoverflow.com/questions/39108742/how-to-get-different-sized-media-twitter-api/39118748

close #773

### CODE REVIEW

**Comment by Tedko** [comment]
Cool. Let’s do a hot fix . And ship it with other ready pr

**Comment by SunriseFox** [comment]
I think using something like `new URL` might be better...

**Comment by Jack-Works** [comment]
> I think using something like `new URL` might be better...

+1 and should add ref into the code

**Comment by guanbinrui** [comment]
@SunriseFox @Jack-Works DONE

But still waiting for bug reporter (a user from tg) tested this build.

File: src/social-network-provider/twitter.com/utils/url.ts [hunk 1]
```diff
@@ -24,6 +24,27 @@ export const getProfileUrl = (self: ProfileIdentifier, isMobile: boolean = false
...
 }
 
+// more about twitter photo url formating: https://developer.twitter.com/en/docs/tweets/data-dictionary/overview/entities-object#photo_format
+export const canonifyImgUrl = (url: string) => {
+    const parsed = new URL(url)
+    if (parsed.hostname !== 'pbs.twimg.com') {
+        return url
+    }
+    const { searchParams } = parsed
+    if (searchParams.get('name') !== 'orig') {
```
Comment (septs) [inline src/social-network-provider/twitter.com/utils/url.ts]: Why not direct set to `"orig"`? orz

Comment (guanbinrui) [reply src/social-network-provider/twitter.com/utils/url.ts]: Fixed! Thank U.

Comment (septs) [reply src/social-network-provider/twitter.com/utils/url.ts]: The condition can be deleted.

File: src/social-network-provider/twitter.com/utils/fetch.ts [hunk 2]
```diff
@@ -172,22 +172,27 @@ export const postImageParser = async (node: HTMLElement) => {
         if (!imgNodes.length) return ''
         const imgUrls = Array.from(imgNodes)
             .filter(node => (!isQuotedTweet ? !node.closest('[role="blockquote"]') : true))
-            .map(node => node.getAttribute('src') ?? '')
+            .map(node => canonifyImgUrl(node.getAttribute('src') ?? ''))
```
Comment (septs) [inline src/social-network-provider/twitter.com/utils/fetch.ts]: Why not `.flatMap(...)`? 呆

Comment (guanbinrui) [reply src/social-network-provider/twitter.com/utils/fetch.ts]: Fixed

Comment (septs) [reply src/social-network-provider/twitter.com/utils/fetch.ts]:
```
.map(node => node.getAttribute('src'))
.filter(src => src) // ignore undefined or empty string
.flatMap(canonifyImgUrl)
```

Comment (guanbinrui) [reply src/social-network-provider/twitter.com/utils/fetch.ts]: It's true. Double map is unnecessary.

File: src/social-network-provider/twitter.com/utils/fetch.ts [hunk 3]
```diff
@@ -171,23 +171,28 @@ export const postImageParser = async (node: HTMLElement) => {
...
+                                    pass: posterIdentity.toText(),
+                                },
+                            )
+                            return /https:\/\/.+\..+\/%20(.+)%40/.test(content) ? content : ''
+                        } catch {
+                            // for twitter image url maybe absent
+                            return ''
+                        }
                     })
                     .filter(Boolean),
```
Comment (septs) [inline src/social-network-provider/twitter.com/utils/fetch.ts]: The `.filter` not working.

```
> [1, 2, 3].map(async (key) => key)
[ Promise { 1 }, Promise { 2 }, Promise { 3 } ]
```

**Pull request merged**